File: demos/01-essentials/01-ai-foundry/05-rag/rag-azure-cs/DocumentProcessor.cs
using System.Text.Json;
using Azure.AI.Inference;
using Azure.AI.OpenAI;
using Azure.Storage.Blobs;

namespace RagAzure;

public class DocumentProcessor
{
    private readonly AzureConfiguration _config;

    public DocumentProcessor(AzureConfiguration config)
    {
        _config = config;
    }

    public async Task<List<SearchDocument>> RetrieveAndProcessDocumentsAsync(
        BlobServiceClient blobServiceClient,
        object embeddingsClient,
        string containerName,
        string blobName)
    {
        Console.WriteLine("\n## 4. Document Retrieval and Processing");
        Console.WriteLine("✅ Document processors initialized");
        
        Console.WriteLine("\n## 5. Retrieve and Process Documents");
        Console.WriteLine(new string('=', 60));
        Console.WriteLine("📥 RETRIEVING PROCESSED DOCUMENTS FROM BLOB STORAGE");
        Console.WriteLine(new string('=', 60));
        
        var containerClient = blobServiceClient.GetBlobContainerClient(containerName);
        var blobClient = containerClient.GetBlobClient(blobName);
        
        if (!await blobClient.ExistsAsync())
        {
            Console.WriteLine($"❌ Blob not found: {blobName}");
            return new List<SearchDocument>();
        }
        
        var download = await blobClient.DownloadContentAsync();
        var jsonContent = download.Value.Content.ToString();
        
        var processedDocuments = JsonSerializer.Deserialize<ProcessedDocumentsWrapper>(jsonContent);
        
        if (processedDocuments?.Policies == null || processedDocuments.Policies.Count == 0)
        {
            Console.WriteLine("❌ No policy documents were found in the processed dataset.");
            return new List<SearchDocument>();
        }
        
        Console.WriteLine($"\n🎉 SUCCESS! Retrieved processed documents from blob storage");
        Console.WriteLine($"📄 Found {processedDocuments.Policies.Count} policy documents");
        
        var chunker = new TextChunker(_config.ChunkSize, _config.ChunkOverlap);
        var searchDocuments = new List<SearchDocument>();
        
        Console.WriteLine($"\n📂 Processing policies documents...");
        var successfulDocs = processedDocuments.Policies.Where(d => d.Success).ToList();
        Console.WriteLine($"✅ Processing {successfulDocs.Count} successful policies documents");
        
        foreach (var doc in successfulDocs)
        {
            if (string.IsNullOrWhiteSpace(doc.Text))
            {
                Console.WriteLine($"⚠️ Skipping document with no text content: {doc.Metadata?.FileName ?? "Unknown"}");
                continue;
            }
            
            var metadata = new Dictionary<string, string>
            {
                ["category"] = "policies",
                ["file_name"] = doc.Metadata?.FileName ?? "Unknown",
                ["file_type"] = doc.Metadata?.FileType ?? "markdown"
            };
            
            var chunks = chunker.ChunkTextForSearch(doc.Text, metadata);
            
            foreach (var chunk in chunks)
            {
                try
                {
                    var embedding = await GenerateEmbeddingAsync(embeddingsClient, chunk.Content);
                    
                    var searchDoc = new SearchDocument
                    {
                        Id = Guid.NewGuid().ToString(),
                        Title = $"{metadata["file_name"]} - Part {chunk.ChunkId + 1}",
                        Content = chunk.Content,
                        Category = "policies",
                        FileName = metadata["file_name"],
                        FileType = metadata["file_type"],
                        ChunkId = chunk.ChunkId,
                        ChunkCount = chunk.ChunkCount,
                        OriginalLength = doc.Text.Length,
                        ChunkLength = chunk.Content.Length,
                        ProcessingDate = DateTimeOffset.UtcNow,
                        ContentVector = embedding
                    };
                    
                    searchDocuments.Add(searchDoc);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"⚠️ Failed to generate embedding for chunk {chunk.ChunkId} of {metadata["file_name"]}: {ex.Message}");
                }
            }
        }
        
        Console.WriteLine($"\n✅ Prepared {searchDocuments.Count} policy document chunks for search index");
        
        if (searchDocuments.Count > 0)
        {
            var totalFiles = searchDocuments.Select(d => d.FileName).Distinct().Count();
            var totalChunks = searchDocuments.Count;
            var avgChunkLength = searchDocuments.Average(d => d.ChunkLength);
            
            Console.WriteLine($"\n📊 POLICIES INDEXING SUMMARY:");
            Console.WriteLine($"   📄 Total policy files: {totalFiles}");
            Console.WriteLine($"   🗂️ Total chunks created: {totalChunks}");
            Console.WriteLine($"   📏 Average chunk length: {avgChunkLength:F0} characters");
            
            var fileStats = searchDocuments.GroupBy(d => d.FileName)
                .Select(g => new { FileName = g.Key, Count = g.Count() });
            
            Console.WriteLine($"\n📋 Policy files breakdown:");
            foreach (var stat in fileStats)
            {
                Console.WriteLine($"   • {stat.FileName}: {stat.Count} chunks");
            }
        }
        
        return searchDocuments;
    }

    private async Task<float[]> GenerateEmbeddingAsync(object client, string text)
    {
        if (client is AzureOpenAIClient openAIClient)
        {
            var embeddingClient = openAIClient.GetEmbeddingClient(_config.EmbeddingsModel);
            var result = await embeddingClient.GenerateEmbeddingAsync(text);
            return result.Value.ToFloats().ToArray();
        }
        else if (client is EmbeddingsClient inferenceClient)
        {
            var embeddingsOptions = new EmbeddingsOptions(new[] { text });
            var result = await inferenceClient.EmbedAsync(embeddingsOptions);
            // The Embedding is BinaryData that needs to be deserialized to float[]
            var embeddingData = result.Value.Data[0].Embedding;
            return embeddingData.ToObjectFromJson<float[]>() ?? Array.Empty<float>();
        }
        else
        {
            throw new InvalidOperationException("Unknown embeddings client type");
        }
    }
}

public class ProcessedDocumentsWrapper
{
    public List<ProcessedDocument> Policies { get; set; } = new();
}

public class ProcessedDocument
{
    public string Text { get; set; } = string.Empty;
    public bool Success { get; set; }
    public DocumentMetadata? Metadata { get; set; }
}

public class DocumentMetadata
{
    public string FileName { get; set; } = string.Empty;
    public string FileType { get; set; } = string.Empty;
}

public class SearchDocument
{
    public string Id { get; set; } = string.Empty;
    public string Title { get; set; } = string.Empty;
    public string Content { get; set; } = string.Empty;
    public string Category { get; set; } = string.Empty;
    public string FileName { get; set; } = string.Empty;
    public string FileType { get; set; } = string.Empty;
    public int ChunkId { get; set; }
    public int ChunkCount { get; set; }
    public int OriginalLength { get; set; }
    public int ChunkLength { get; set; }
    public DateTimeOffset ProcessingDate { get; set; }
    public float[] ContentVector { get; set; } = Array.Empty<float>();
}

File: demos/01-essentials/02-agentic-ai/02-rag/rag-azure-py/document_processor.py
import uuid
from datetime import datetime
from typing import List, Dict
from tqdm import tqdm
from document_retriever import DocumentRetriever
from text_chunker import TextChunker


def retrieve_and_process_documents(
    blob_service_client,
    embeddings_client,
    container_name: str,
    blob_name: str,
    chunk_size: int,
    chunk_overlap: int,
    embeddings_model: str
) -> List[Dict]:
    retriever = DocumentRetriever(blob_service_client, container_name, blob_name)
    chunker = TextChunker(
        chunk_size=chunk_size,
        chunk_overlap=chunk_overlap
    )
    
    print("\n" + "="*60)
    print("📥 RETRIEVING PROCESSED DOCUMENTS FROM BLOB STORAGE")
    print("="*60)
    
    processed_documents = retriever.get_all_processed_documents()
    if not processed_documents:
        print("❌ No processed documents available. Ensure the preprocessing notebook has been run.")
        return []
    
    print(f"\n🎉 SUCCESS! Retrieved processed documents from blob storage")
    print(f"📊 Available categories: {list(processed_documents.keys())}")
    
    policies_only = {'policies': processed_documents.get('policies', [])}
    if not policies_only['policies']:
        print("❌ No policy documents were found in the processed dataset.")
        return []
    
    print(f"🎯 Filtering to process POLICIES only...")
    print(f"📄 Found {len(policies_only['policies'])} policy documents")
    
    search_documents = []
    
    for category, docs in policies_only.items():
        print(f"\n📂 Processing {category} documents...")
        
        successful_docs = [doc for doc in docs if doc.get('success', False)]
        print(f"✅ Processing {len(successful_docs)} successful {category} documents")
        
        for doc in tqdm(successful_docs, desc=f"Processing {category}"):
            text_content = doc.get('text', '')
            if not text_content:
                print(f"⚠️ Skipping document with no text content: {doc.get('metadata', {}).get('file_name', 'Unknown')}")
                continue
            
            metadata = doc.get('metadata', {}).copy()
            metadata['category'] = category
            
            chunks = chunker.chunk_text_for_search(text_content, metadata)
            
            for chunk in chunks:
                try:
                    embedding_response = embeddings_client.embed(
                        input=[chunk['content']],
                        model=embeddings_model,
                    )
                    content_vector = embedding_response.data[0].embedding
                except Exception as error:  # noqa: BLE001
                    print(
                        f"⚠️ Failed to generate embedding for chunk {chunk['chunk_id']} "
                        f"of {metadata.get('file_name', 'Unknown')}: {error}"
                    )
                    continue
                
                search_doc = {
                    'id': str(uuid.uuid4()),
                    'title': f"{metadata.get('file_name', 'Unknown')} - Part {chunk['chunk_id'] + 1}",
                    'content': chunk['content'],
                    'category': category,
                    'file_name': metadata.get('file_name', 'Unknown'),
                    'file_type': metadata.get('file_type', 'markdown'),
                    'chunk_id': chunk['chunk_id'],
                    'chunk_count': chunk['chunk_count'],
                    'original_length': len(text_content),
                    'chunk_length': len(chunk['content']),
                    'processing_date': datetime.now().isoformat() + 'Z',
                    'content_vector': content_vector
                }
                search_documents.append(search_doc)
    
    print(f"\n✅ Prepared {len(search_documents)} policy document chunks for search index")
    
    if search_documents:
        total_files = len(set(doc['file_name'] for doc in search_documents))
        total_chunks = len(search_documents)
        avg_chunk_length = sum(doc['chunk_length'] for doc in search_documents) / total_chunks
        
        print(f"\n📊 POLICIES INDEXING SUMMARY:")
        print(f"   📄 Total policy files: {total_files}")
        print(f"   🗂️ Total chunks created: {total_chunks}")
        print(f"   📏 Average chunk length: {avg_chunk_length:.0f} characters")
        
        file_stats = {}
        for doc in search_documents:
            file_name = doc['file_name']
            if file_name not in file_stats:
                file_stats[file_name] = 0
            file_stats[file_name] += 1
        
        print(f"\n📋 Policy files breakdown:")
        for file_name, chunk_count in file_stats.items():
            print(f"   • {file_name}: {chunk_count} chunks")
    
    return search_documents
